app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx

```diff
@@ -1,29 +1,204 @@
-    console.log("tracks", props.pdb.tracks);
-
-    const svgRef = React.useRef<SVGSVGElement>(null);
+const margin = {
+    top: 10,
+    right: 10,
+    bottom: 10,
+    left: 20,
+};
 
-    React.useEffect(() => {
-        if (!svgRef.current) return;
+const dimensions = {
+    width: 200 - margin.left - margin.right,
+    height: 200 - margin.top - margin.bottom,
+};
 
-        const svg = d3.select(svgRef.current);
-        svg.append("circle").attr("cx", 140).attr("cy", 70).attr("r", 40).style("fill", "red");
-        svg.append("circle").attr("cx", 300).attr("cy", 100).attr("r", 40).style("fill", "green");
-    }, []);
+export const ProtvistaPdbValidation: React.FC<ProtvistaPdbProps> = React.memo(props => {
+    const ref = useGrid();
 
     return (
         <>
             <h2>This is the custom ProtvistaPdbValidation component</h2>
-            <svg ref={svgRef} width={5000} height={200} />
+            <svg ref={ref} />
             <ProtvistaPdb {...props} />
         </>
     );
 });
+
+function useGrid() {
+    const svgRef = React.useRef<SVGSVGElement>(null);
+
+    React.useEffect(() => {
+        if (!svgRef.current) return;
+
+        let i = 0;
+        const numRows = modelQualityStats.graph.axis_y.categories.length;
+        const numCols = modelQualityStats.graph.axis_x.categories.length;
+        const axisX = modelQualityStats.graph.axis_x.axis_name;
+        const axisY = modelQualityStats.graph.axis_y.axis_name;
+
+        const gridData = () => {
+            const data: any[] = [];
+            const width = 40;
+            let height = 40;
+            let xPos = 1;
+            let yPos = height * numRows;
+            for (let row = 0; row < numRows; row++) {
+                data.push([]);
+                for (let column = 0; column < numCols; column++) {
+                    const cellXValue = modelQualityStats.data.categories[i]?.cat_id.split(",")[0];
+                    const cellYValue = modelQualityStats.data.categories[i]?.cat_id.split(",")[1];
+                    data[row].push({
+                        x: xPos,
+                        y: yPos,
+                        width: width,
+                        height: height,
+                        value: modelQualityStats.data.categories[i]?.count,
+                        color:
+                            (cellXValue === "1" || cellXValue === "3") &&
+                            (cellYValue === "1" || cellYValue === "2")
+                                ? "red"
+                                : (cellXValue === "1" || cellXValue === "3") && cellYValue === "3"
+                                ? "orange"
+                                : cellXValue === "2" && cellYValue === "4"
+                                ? "lime"
+                                : cellXValue === "2" && cellYValue === "3"
+                                ? "greenyellow"
+                                : "yellow",
```

Comment:
Long object value. Create a local variable / separate function.

Reply:
i'm adding the colours to the JSON file so i won't be using this anymore.